### PR TITLE
parser: sanitize timestamps to RFC3339

### PR DESCRIPTION
**Pull request by mdibaiee**
**Description:**

- Automatically parse and sanitize timestamps to RFC3339

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

### CODE REVIEW

**Review by psFried — CHANGES_REQUESTED** [review]
I think my main feedback is that this needs more exhaustive test cases, particularly for different numbers of sub-second digits and different representations of timezone offsets.

I left some other comments and questions. Let me know if it'd help to talk through any of those.

File: crates/parser/src/format/sanitize/datetime.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+const NAIVE_FORMATS: [&'static str; 4] = [
+    "%Y-%m-%dT%H:%M:%S",
+    "%Y-%m-%dT%H:%M:%S%.3f",
+    "%Y-%m-%d %H:%M:%S%.3f",
+    "%Y-%m-%d %H:%M:%S",
+];
+
+const FORMATS: [&'static str; 2] = [
+    "%Y-%m-%d %H:%M:%S%.3f%:z",
```
[the remark this answers is not in the document]
Comment (mdibaiee) [reply crates/parser/src/format/sanitize/datetime.rs:19]: > Another thing is that I don't see the T here. I see that some test cases have it, so I'm wondering if those cases are passing because we're just passing them through as is. Is that intentional and important? If so, it's deserving of a comment.

Yes we just pass through RFC3339 as is, it doesn;t need us to parse it and format it again

[the remark this answers is not in the document]
Comment (mdibaiee) [reply crates/parser/src/format/sanitize/datetime.rs:19]: > I also wonder if chrono's built-in rfc3339 format would handle all of those cases without needing to try both of these formats. Is there a reason not to use that?

RFC3339 requires the `T`, here we are trying to handle the case where the `T` is missing, which is not valid RFC3339

**Comment by mdibaiee** [comment]
Okay, so I did a few things here:

1. First, I added a test case with fractional seconds up to nanosecond precision (9 digits)
2. I added a simple benchmark for us to use to compare the sanitization with no sanitization
3. I ran the benchmark with sanitization enabled and disabled, and the result was that with sanitization, it is approximately %11-13 slower on this dataset:
```
peoples_500             time:   [5.0192 ms 5.0818 ms 5.1640 ms]
                        change: [+11.122% +12.920% +14.831%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) high mild
  13 (13.00%) high severe
```

4. I switched from `chrono` to `time` crate, this makes the code simpler, but in terms of performance it didn't really help. It is still ~12% slower.

I'll look into this more tomorrow to see if I can improve performance somehow.

**Comment by mdibaiee** [comment]
@psFried made a change now, with this change:

1. we first attempt to parse into a more _generic_ format that is a superset of naive and offset datetimes
2. if this parsing succeeds, when then try to parse into an offset date
3. if we succeed in parsing an offset date, we emit this datetime and we are done
4. if we don't succeed in parsing an offset date, then this is a naive date and we assume the default offset for it and then emit

This makes the performance impact of sanitization ~4%

File: crates/parser/src/format/sanitize/datetime.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+struct DatetimeSanitizer {
+    from: Output,
+    default_offset: time::UtcOffset,
+}
+
+// Here we are trying to parse non-RFC3339 dates
+fn datetime_to_rfc3339(val: &mut Value, default_offset: time::UtcOffset) {
+    match val {
+        Value::String(s) => {
+            let primitive_format = format_description!(
```
Comment (psFried) [inline crates/parser/src/format/sanitize/datetime.rs:14]: nit: I think this could use some comments explaining why we're using 2 formats and the purpose of each one.

**Review by psFried — CHANGES_REQUESTED** [review]
Some other edge cases that I think need test coverage are inputs that have explicit TZ offsets that aren't the same as the default: for example, `2023-09-26 12:34:56-04:00` should get normalized to `2023-09-26T12:34:56-04:00`, and `2023-09-26T12:34:56-04:00` should be passed through unaltered.

Also, the existing test cases for fractional seconds are all for zeros, which get truncated by the normalization process. I think we should have some tests for non-zero fractional seconds that assert that the fractional portions don't lose significant digits.

**Review by psFried — APPROVED** [review]
Those new test cases are way easier to read and understand. Thanks! I think this is looking in nice shape.
